src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -274,10 +275,28 @@ async def mypy_typecheck_partition(
 
                             {mkdir.path} -p {run_cache_dir}/{py_version} > /dev/null 2>&1 || true
                             {cp.path} {mypy_cache_dir}/{py_version}/cache.db {run_cache_dir}/{py_version}/cache.db > /dev/null 2>&1 || true
+
                             {' '.join((shell_quote(arg) for arg in argv))}
                             EXIT_CODE=$?
+
+                            get_device_id() {{
+                                TEST_PATH=$({readlink.path} -f "$1")
+                                stat -c "%d" "$TEST_PATH" 2>/dev/null || stat -f "%d" "$TEST_PATH" 2>/dev/null
```

Comment:
So this is `-c` for linux, `-f` for macos? Worth documenting, I reckon.

Also, dammit BSD for not getting with the progrma.

Reply:
So, this is the wrong test anyway. A mv is only atomic on the same filesystem. A given device can have multiple filesystems mounted on it. You might get lucky, but the problem is not solved for real. You might try hardlinking a file in the one spot to the other though, that has the same single filesystem requirement as an atomic rename.